client/api/personalization/use-user-preferences.ts

```diff
@@ -14,54 +14,65 @@ type UserPreferencesResponse = {
 
 const GET_USER_PREFERENCES_QUERY = 'GET_USER_PREFERENCES_QUERY';
 
-function getUserPreferences(fpData: GetResult): Promise<UserPreferencesResponse> {
-  return apiRequest('/api/personalization/get-user-preferences', fpData);
+function getUserPreferencesFromServer(fpData: GetResult, abortSignal: AbortSignal): Promise<UserPreferencesResponse> {
+  return apiRequest('/api/personalization/get-user-preferences', fpData, {}, abortSignal);
 }
 
-function updateUserPreferences(fpData: GetResult, preferences: UserPreferences) {
+function updateUserPreferencesOnServer(fpData: GetResult, preferences: UserPreferences) {
   return apiRequest('/api/personalization/update-user-preferences', fpData, preferences);
 }
 
 export function useUserPreferences() {
   // Get visitorId
   const { data: fingerprintResult } = useVisitorData();
 
+  // An abort controller to cancel the query if the user changes dark mode while the query is still loading
+  const abortController = useMemo(() => new AbortController(), []);
+
   // Query database for user preferences for this visitorId
-  const userPreferencesQuery = useQuery(GET_USER_PREFERENCES_QUERY, () => getUserPreferences(fingerprintResult), {
-    enabled: Boolean(fingerprintResult),
-    onSuccess: (data) => {
-      // Store the result in localStorage to get it faster on page reload
-      localStorage.setItem('hasDarkMode', String(data?.data?.hasDarkMode));
-    },
-  });
+  const { data: preferencesResponse } = useQuery(
+    GET_USER_PREFERENCES_QUERY,
+    () => getUserPreferencesFromServer(fingerprintResult, abortController.signal),
+    {
+      enabled: Boolean(fingerprintResult),
+      onSuccess: (data) => {
+        // Store the result in localStorage to get it faster on page reload
+        localStorage.setItem('hasDarkMode', String(data?.data?.hasDarkMode));
+      },
+    }
+  );
 
   // Use the same query object to update preferences synchronously further down
   const queryClient = useQueryClient();
-  const updateUserPreferencesData = useCallback(
-    (data: UserPreferences) => queryClient.setQueryData<UserPreferencesResponse>(GET_USER_PREFERENCES_QUERY, { data }),
+  const updateUserPreferencesOnClient = useCallback(
+    (data: UserPreferences) => {
+      queryClient.setQueryData<UserPreferencesResponse>(GET_USER_PREFERENCES_QUERY, { data });
+    },
     [queryClient]
   );
 
   // On component mount, set the global state of the user preferences to the value stored in localStorage
   // (to prevent a long flash of light mode while waiting for userPreferencesQuery)
   useEffect(() => {
-    updateUserPreferencesData({ hasDarkMode: localStorage.getItem('hasDarkMode') === 'true' });
-  }, [updateUserPreferencesData]);
+    updateUserPreferencesOnClient({ hasDarkMode: localStorage.getItem('hasDarkMode') === 'true' });
+  }, [updateUserPreferencesOnClient]);
```

Comment:
You could try to use [placeholder](https://tanstack.com/query/v4/docs/react/guides/placeholder-query-data) data to pass it to query on mount

Reply:
Good idea, but I am not sure how to access localStorage outside of a useEffect: 
<img width="1017" alt="image" src="https://github.com/fingerprintjs/fingerprintjs-pro-use-cases/assets/11249993/70c11c1f-ed08-42cd-9663-dfac48d05a66">
 
Any ideas?

Reply:
Could you try to do `localStorage?.getItem` and see if it works?
The error is probably caused by SSR, if my suggestion won't work let's just keep it as it is

Reply:
So interestingly both `localStorage?.getItem()` and `localStorage && localStorage.getItem()` does not work. 

What works is `typeof localStorage !== 'undefined' && localStorage.getItem()` but it does not behave like we would want here. Setting the `initialData` does not force a re-render of the components that use dark more, so the they only turn black once they re-render naturally :D 
![image](https://github.com/fingerprintjs/fingerprintjs-pro-use-cases/assets/11249993/24b9a32f-f01a-4f53-8d02-d906160e9708)

Is there a way to force the re-render? If not, the original useEffect might not be so bad

Reply:
Okay :)